### PR TITLE
implemented d.ts for actions

### PR DESCRIPTION
**Pull request by munkk**

### CODE REVIEW

File: src/core/infrastructure/cache.d.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export declare class Cache {
```
Comment (vkorolev) [inline src/core/infrastructure/cache.d.ts]: Can we make it generic like `class Cache<K, V>`?

Comment (klumba12) [reply src/core/infrastructure/cache.d.ts]: I would say yes =)

File: src/core/action/action.model.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+declare class ActionModel{
+	constructor();
+
+	items: Array<any>;
```
Comment (vkorolev) [inline src/core/action/action.model.d.ts]: Consider to make it `Array<Action>` and necessary types

File: src/core/infrastructure/cache.d.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+	get(key: K): any ;
+
+	has(key: K): boolean;
+
+	find(key: K): any;
+
+	remove(key: K): void;
+
+	clear(): void;
+}
```
Comment (munkk) [inline src/core/infrastructure/cache.d.ts:16]: did i fix this right?

Comment (vkorolev) [reply src/core/infrastructure/cache.d.ts:16]: Almost:)
`get` and `find` should return `V` instead of any

Comment (munkk) [reply src/core/infrastructure/cache.d.ts:16]: thx)

File: src/core/action/action.model.d.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import {Action} from '../infrastructure/action';
```
Comment (vkorolev) [inline src/core/action/action.model.d.ts:1]: Looks like you forgot to include Action file

File: src/core/infrastructure/command.d.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+export declare class Command {
+   constructor(context: Object );
```
Comment (vkorolev) [inline src/core/infrastructure/command.d.ts]: context, execute, canExecute and shortcut should have types too

File: src/core/infrastructure/action.d.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+export declare class Action {
+
+	constructor(command: Object, title: Object, icon: Object);
+
+	command: Object;
```
Comment (klumba12) [inline src/core/infrastructure/action.d.ts]: Should be Command Type

File: src/core/infrastructure/action.d.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+	title: Object;
```
Comment (klumba12) [inline src/core/infrastructure/action.d.ts]: title & icon -> string

File: src/core/infrastructure/error.d.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+export declare class AppError{
+
+	constructor(name: String, message: String);
+
+	name: String;
```
Comment (klumba12) [inline src/core/infrastructure/error.d.ts]: I guess no need to declare name and message AppError used only as exception source

File: src/core/infrastructure/event.d.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+	constructor(e: Object) ;
+
+	handlers: Array<any>
+	isDirty: Boolean;
+	e: Object;
+
+
+	on(f: Object): void;
+
+	watch(f: Object): void
```
Comment (klumba12) [inline src/core/infrastructure/event.d.ts]: f: should be a function type

File: src/core/infrastructure/event.d.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+export declare class Event {
+
+	constructor(e: Object) ;
+
+	handlers: Array<any>
+	isDirty: Boolean;
+	e: Object;
```
Comment (klumba12) [inline src/core/infrastructure/event.d.ts]: no need to declare handlers isDirty and e - they are private here

File: src/core/infrastructure/event.listener.d.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
+export declare class EventListener {
+
+	constructor(context: Object, element: Object);
+
+	element: Object;
+	context: Object;
+	handlers: Object;
```
Comment (klumba12) [inline src/core/infrastructure/event.listener.d.ts]: these 3 are private

File: src/core/infrastructure/event.listener.d.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+
+	on(name: String, f: Object): void;
```
Comment (klumba12) [inline src/core/infrastructure/event.listener.d.ts]: f is function

File: src/core/infrastructure/fetch.d.ts [hunk 13]
```diff
@@ -0,0 +1,9 @@
+export declare class Fetch {
+	constructor(select: Object);
+
+	select: Object;
+	busy: Boolean;
+	result: Boolean;
```
Comment (klumba12) [inline src/core/infrastructure/fetch.d.ts]: I'm not sure if these 2 are boolean

File: src/core/infrastructure/shortcut.d.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
+export declare class Shortcut {
+
+	constructor(table: Object, manager: Object);
+
+	manager: Object;
+	commands: Array<any>;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: Array of Command Types

File: src/core/infrastructure/fetch.d.ts [hunk 15]
```diff
@@ -0,0 +1,9 @@
+export declare class Fetch {
+	constructor(select: Object);
+
+	select: Object;
+	busy: null;
```
Comment (klumba12) [inline src/core/infrastructure/fetch.d.ts]: null is not a type :)

File: src/core/infrastructure/event.d.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+export declare class Event {
+
+	constructor(e: object) ;
+
+	on(f: object): void;
+
+	watch(f: object): void
+
+	emit(e: object): void;
```
Comment (klumba12) [inline src/core/infrastructure/event.d.ts]: f - should be a function not a object

File: src/core/infrastructure/event.listener.d.ts [hunk 17]
```diff
@@ -0,0 +1,13 @@
+export declare class EventListener {
+
+	constructor(context: object, element: object);
+
+	private element: object;
+	private context: object;
+	private handlers: object;
```
Comment (klumba12) [inline src/core/infrastructure/event.listener.d.ts]: private properties shouldn't be in declaration

File: src/core/infrastructure/fetch.d.ts [hunk 18]
```diff
@@ -0,0 +1,9 @@
+export declare class Fetch {
+	constructor(select: Object);
```
Comment (klumba12) [inline src/core/infrastructure/fetch.d.ts]: select should be "any"

File: src/core/infrastructure/fetch.d.ts [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+
+	select: Object;
+	busy: null;
+	result: null;
```
Comment (klumba12) [inline src/core/infrastructure/fetch.d.ts]: result is any

File: src/core/infrastructure/model.bind.d.ts [hunk 20]
```diff
@@ -0,0 +1,8 @@
+export class ModelBinder {
+	constructor(source: Object);
+
+	source: Object;
+	off: Object;
+
+	bind(model: Object, names: Array, run: Boolean): Object;
```
Comment (klumba12) [inline src/core/infrastructure/model.bind.d.ts]: names: Array of strings

File: src/core/infrastructure/shortcut.d.ts [hunk 21]
```diff
@@ -0,0 +1,26 @@
+import {Command} from './command';
+
+export declare class Shortcut {
+
+	constructor(table: object, manager: object);
+
+	manager: object;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: manager should be CommandManager you can put decloration of CommandManager in infrastructure folder near shortcut

File: src/core/infrastructure/shortcut.d.ts [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+	commands: Command[];
+	shortcuts: object;
+	codeMap: object;
+   canExecute : object;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: function

File: src/core/infrastructure/shortcut.d.ts [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+	off: object;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: function

Comment (klumba12) [reply src/core/infrastructure/shortcut.d.ts]: I'm not sure that we have this

File: src/core/infrastructure/shortcut.d.ts [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+
+
+	translate(e: object): any[];
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: is there declaration for DomEvent?

File: src/core/infrastructure/shortcut.d.ts [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+
+	onKeyDown(e: object): void;
+
+	register(id: number, commands: object): any;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: register returns function?

File: src/core/infrastructure/shortcut.d.ts [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+
+	find(code: string): any[];
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: this is private, remove it

File: src/core/infrastructure/shortcut.d.ts [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+
+	test(shortcut: object, code: string): any;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: this is private do not include too

File: src/core/services/convert.d.ts [hunk 28]
```diff
@@ -0,0 +1,16 @@
+export declare function parseFactory (type: string): any;
+
+export declare function getType(value: any): string;
+
+export declare function isPrimitive(type: string): boolean;
+
+export declare function parseBool(value: string): boolean;
```
Comment (klumba12) [inline src/core/services/convert.d.ts]: i'm not sure that for all parse methods value is not any?

File: src/core/services/dom.d.ts [hunk 29]
```diff
@@ -0,0 +1,5 @@
+export declare function css(element: object, property: string, value: string): any;
```
Comment (klumba12) [inline src/core/services/dom.d.ts]: element maybe should be DomElement or HtmlElement do we have that is standart definitions?

File: src/core/services/dom.observe.d.ts [hunk 30]
```diff
@@ -0,0 +1,5 @@
+export declare function children(element: object): object;
```
Comment (klumba12) [inline src/core/services/dom.observe.d.ts]: DomElements not object, also sometimes Object is used sometimes object choose one please

File: src/core/services/grid.d.ts [hunk 31]
```diff
@@ -0,0 +1,9 @@
+
+export class GridService {
+
+	constructor(model: object, apply: object);
```
Comment (klumba12) [inline src/core/services/grid.d.ts]: apply is function model is Model

File: src/core/services/label.d.ts [hunk 32]
```diff
@@ -0,0 +1,5 @@
+export declare function get(row: object, column: object): any;
```
Comment (klumba12) [inline src/core/services/label.d.ts]: column is ColumnModel

File: src/core/services/merge.d.ts [hunk 33]
```diff
@@ -0,0 +1 @@
+﻿export declare function merge(settings): any;
```
Comment (klumba12) [inline src/core/services/merge.d.ts]: make definition for settings too can be done in the same file

File: src/core/services/middleware.d.ts [hunk 34]
```diff
@@ -0,0 +1,8 @@
+export declare class Middleware {
+
+	constructor();
+
+	run(context: object, memo: any[]): any;
```
Comment (klumba12) [inline src/core/services/middleware.d.ts]: Make definition for context please can be done in the same file

File: src/core/services/pair.d.ts [hunk 35]
```diff
@@ -0,0 +1,9 @@
+export declare function key(pair: object): string;
```
Comment (klumba12) [inline src/core/services/pair.d.ts]: investigate what we can use here instead of object

File: src/core/services/path.d.ts [hunk 36]
```diff
@@ -0,0 +1,2 @@
+export declare function compile(path: string): any;
```
Comment (klumba12) [inline src/core/services/path.d.ts]: return specific function

File: src/core/services/value.d.ts [hunk 37]
```diff
@@ -0,0 +1,5 @@
+export declare function get(row: object, column: object): any;
```
Comment (klumba12) [inline src/core/services/value.d.ts]: column should be Column

File: src/core/infrastructure/fetch.d.ts (hunk 15 above)
Comment (titovmx) [reply src/core/infrastructure/fetch.d.ts]: actually ```busy``` is a Promise

File: src/core/services/convert.d.ts (hunk 28 above)
Comment (titovmx) [reply src/core/services/convert.d.ts]: In my opinion parse methods should return target type exactly

File: src/core/services/convert.d.ts [hunk 38]
```diff
@@ -0,0 +1,16 @@
+export declare function parseFactory (type: string): any;
```
Comment (titovmx) [inline src/core/services/convert.d.ts]: Should ```parseFactory``` return ```Function```?

File: src/core/services/css.d.ts [hunk 39]
```diff
@@ -0,0 +1,5 @@
+export declare function sheet(id: string): any;
+
+export declare function sheet(id: string): any;
```
Comment (titovmx) [inline src/core/services/css.d.ts]: ```sheet``` is duplicated

File: src/core/services/label.d.ts [hunk 40]
```diff
@@ -0,0 +1,5 @@
...
+
+export function getFactory(column: object): any;
```
Comment (titovmx) [inline src/core/services/label.d.ts]: returns ```Function```

File: src/core/services/value.d.ts [hunk 41]
```diff
@@ -0,0 +1,5 @@
...
+
+export declare function getFactory(column: object): any;
```
Comment (titovmx) [inline src/core/services/value.d.ts]: returns ```Function``` too

File: src/core/infrastructure/event.listener.d.ts [hunk 42]
```diff
@@ -0,0 +1,13 @@
+export declare class EventListener {
+
+	constructor(context: object, element: object);
+
+	element: object;
+	context: object;
+	handlers: object;
```
Comment (klumba12) [inline src/core/infrastructure/event.listener.d.ts]: remove element, context, handlers please

File: src/core/infrastructure/fetch.d.ts [hunk 43]
```diff
@@ -0,0 +1,10 @@
+export declare class Fetch {
+
+	constructor(select: any);
+
+	select: any;
+	busy: object;
```
Comment (klumba12) [inline src/core/infrastructure/fetch.d.ts]: busy is Promise

File: src/core/infrastructure/fetch.d.ts [hunk 44]
```diff
@@ -0,0 +1,10 @@
...
+	result: any;
+
+	run(item: object): any;
```
Comment (klumba12) [inline src/core/infrastructure/fetch.d.ts]: item is any i guess...

File: src/core/infrastructure/model.bind.d.ts [hunk 45]
```diff
@@ -0,0 +1,8 @@
+export class ModelBinder {
+	constructor(source: object);
+
+	source: object;
```
Comment (klumba12) [inline src/core/infrastructure/model.bind.d.ts]: remove source

File: src/core/infrastructure/model.bind.d.ts [hunk 46]
```diff
@@ -0,0 +1,8 @@
...
+	off: object;
+
+	bind(model: object, names: string[], run: boolean): object;
```
Comment (klumba12) [inline src/core/infrastructure/model.bind.d.ts]: returns not ojbect but function

File: src/core/infrastructure/model.d.ts [hunk 47]
```diff
@@ -0,0 +1,5 @@
+export declare class Model {
+	constructor();
+
+	register(name: string, model: object): object;
```
Comment (klumba12) [inline src/core/infrastructure/model.d.ts]: register returns Model itself

File: src/core/infrastructure/shortcut.d.ts [hunk 48]
```diff
@@ -0,0 +1,27 @@
+import {Command} from './command';
+import {CommandManager} from 'view/services/command.manager';
+
+export declare class Shortcut {
+
+	constructor(table: object, manager: CommandManager);
+
+	manager: CommandManager;
+	commands: Command[];
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts:9]: sometimes u use Array<> sometimest [], please make it the same everywhere

File: src/core/infrastructure/shortcut.d.ts [hunk 49]
```diff
@@ -0,0 +1,27 @@
...
+	shortcuts: () => any;
+	codeMap: () => any;
+   canExecute: () => boolean;
+	off: object;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: remove off, shortucts, codeMap and canExecute they are private

File: src/core/infrastructure/shortcut.d.ts [hunk 50]
```diff
@@ -0,0 +1,27 @@
...
+
+
+	translate(e: object): any[];
+
+	onKeyDown(e: object): void;
```
Comment (klumba12) [inline src/core/infrastructure/shortcut.d.ts]: remove translate and onKeyDown

File: src/core/infrastructure/command.d.ts [hunk 51]
```diff
@@ -0,0 +1,7 @@
+export declare class Command {
+   constructor(context: object);
+
+   execute(): void;
```
Comment (klumba12) [inline src/core/infrastructure/command.d.ts]: please fix formatting, also execute and canExecute can accept arguments, is'it possible to declare that?

File: src/core/services/value.d.ts [hunk 52]
```diff
@@ -0,0 +1,8 @@
+import {ColumnModel} from '../column-type/column.model';
+import {Row} from '../row/row';
+
+export declare function get(row: Row, column: ColumnModel): any;
```
Comment (titovmx) [inline src/core/services/value.d.ts]: I'm not sure about first parameter as ```Row```. It should be one of objects from ```model.data().rows``` which are collection of objects. So I think, ```row``` is ```object``` too

File: src/core/infrastructure/event.d.ts [hunk 53]
```diff
@@ -0,0 +1,10 @@
+export declare class Event {
+
+	constructor(e: () => null) ;
+
+	on(f: () => any): any;
```
Comment (klumba12) [inline src/core/infrastructure/event.d.ts]: on returns function to off :)

File: src/core/infrastructure/model.bind.d.ts [hunk 54]
```diff
@@ -0,0 +1,8 @@
+export class ModelBinder {
+
+	constructor(source: object);
+
+	off: object;
```
Comment (klumba12) [inline src/core/infrastructure/model.bind.d.ts]: off is a function

File: src/core/utility/utility.d.ts [hunk 55]
```diff
@@ -0,0 +1,11 @@
+export declare const noop: object;
```
Comment (klumba12) [inline src/core/utility/utility.d.ts]: noop is function, identiy is function

File: src/core/services/grid.d.ts [hunk 56]
```diff
@@ -0,0 +1,10 @@
+import {Model} from "../infrastructure/model";
+
+export class GridService {
+
+	constructor(model: Model, apply: () => object);
+
+	invalidate(source: string, changes: object, pipe: object): any;
```
Comment (klumba12) [inline src/core/services/grid.d.ts]: pipe should have a type

Comment (klumba12) [reply src/core/services/grid.d.ts]: invalidate returns promise?

File: src/core/utility/utility.d.ts [hunk 57]
```diff
@@ -0,0 +1,11 @@
+export declare const noop: () => Object;
```
Comment (klumba12) [inline src/core/utility/utility.d.ts]: returns void

File: src/core/utility/utility.d.ts [hunk 58]
```diff
@@ -0,0 +1,11 @@
...
+
+export declare const yes: () => boolean;
+
+export declare const no: () => boolean;
+
+export declare const identity: (arg: object) => any;
```
Comment (klumba12) [inline src/core/utility/utility.d.ts]: (any) => any

File: src/core/services/value.d.ts [hunk 59]
```diff
@@ -0,0 +1,7 @@
+import {ColumnModel} from '../column-type/column.model';
+
+export declare function get(row: object, column: ColumnModel): any;
+
+export declare function getFactory(column: ColumnModel): any;
```
Comment (klumba12) [inline src/core/services/value.d.ts]: returns (row: any) => any

File: src/core/services/value.d.ts [hunk 60]
```diff
@@ -0,0 +1,7 @@
...
+
+export function set(row: object, column: ColumnModel, value: string): any;
```
Comment (klumba12) [inline src/core/services/value.d.ts]: returns void

File: src/core/services/path.d.ts [hunk 61]
```diff
@@ -0,0 +1,6 @@
+interface IReturnFunc{
+	(entry: object, value: any): string;
```
Comment (klumba12) [inline src/core/services/path.d.ts]: returns any

File: src/core/services/label.d.ts [hunk 62]
```diff
@@ -0,0 +1,8 @@
+import {ColumnModel} from '../column-type/column.model';
+
+
+export declare function get(row: object, column: ColumnModel): any;
+
+export function getFactory(column: ColumnModel): any;
```
Comment (klumba12) [inline src/core/services/label.d.ts]: returns (row: any) => any

File: src/core/services/label.d.ts [hunk 63]
```diff
@@ -0,0 +1,8 @@
...
+
+export function set(row: object, column: ColumnModel, label: object): any;
```
Comment (klumba12) [inline src/core/services/label.d.ts]: returns void

File: src/core/services/dom.observe.d.ts [hunk 64]
```diff
@@ -0,0 +1,5 @@
+export declare function children(element: HTMLElement): object;
+
+export function style(element: HTMLElement): object;
+
+export function className(element: HTMLElement): object;
```
Comment (klumba12) [inline src/core/services/dom.observe.d.ts]: all funcs returns not object but sepcial interface {on:...}

File: src/core/services/dom.d.ts [hunk 65]
```diff
@@ -0,0 +1,5 @@
+export declare function css(element: HTMLElement, property: string, value: string): any;
```
Comment (klumba12) [inline src/core/services/dom.d.ts]: return string i guess..

File: src/core/services/css.d.ts [hunk 66]
```diff
@@ -0,0 +1,3 @@
...
```
Comment (klumba12) [inline src/core/services/css.d.ts]: Returns HTMLElement?

File: src/core/infrastructure/shortcut.d.ts (hunk 48 above)
Comment (klumba12) [reply src/core/infrastructure/shortcut.d.ts:9]: Up

File: src/core/infrastructure/event.d.ts [hunk 67]
```diff
@@ -0,0 +1,10 @@
+export declare class Event {
+
+	constructor(e: () => null) ;
+
+	on(f: () => any): () => void;
```
Comment (klumba12) [inline src/core/infrastructure/event.d.ts]: If u use interfaces for returned functions do it everywhere please:)

File: src/core/services/merge.d.ts [hunk 68]
```diff
@@ -0,0 +1 @@
+﻿export declare function merge(settings: object): any;
```
Comment (klumba12) [inline src/core/services/merge.d.ts]: mrege return statistics object not any

File: src/core/services/markup.d.ts [hunk 69]
```diff
@@ -0,0 +1 @@
+export declare function build(style: object): object;
```
Comment (klumba12) [inline src/core/services/markup.d.ts]: returns string

File: src/core/services/label.d.ts [hunk 70]
```diff
@@ -0,0 +1,11 @@
+import {ColumnModel} from '../column-type/column.model';
+
+interface IGetResult{
+	(row: any); any;
+}
+
+export declare function get(row: object, column: ColumnModel): IGetResult;
+
+export function getFactory(column: ColumnModel): any;
```
Comment (klumba12) [inline src/core/services/label.d.ts]: this returns IGetResult

File: src/core/services/label.d.ts [hunk 71]
```diff
@@ -0,0 +1,11 @@
+import {ColumnModel} from '../column-type/column.model';
+
+interface IGetResult{
+	(row: any); any;
+}
+
+export declare function get(row: object, column: ColumnModel): IGetResult;
```
Comment (klumba12) [inline src/core/services/label.d.ts]: returns any

File: src/core/services/aggregation.d.ts [hunk 72]
```diff
@@ -0,0 +1,23 @@
...
+
+	static avg(rows: object[], getValue: (param: string) => any, options: any[]) any;
+
+	static sum(rows: object[], getValue: (param: string) => any, options: any[], container: any) any;
+
+	static join(rows: object[], getValue: (param: string) => any, options: any[]) any;
+
+	static count(rows: object[], getValue: (param: string) => any, options: any[]) any;
+
+}
```
Comment (klumba12) [inline src/core/services/aggregation.d.ts]: rename param to row every-where,
options should have type too :)

File: src/core/infrastructure/command.d.ts [hunk 73]
```diff
@@ -0,0 +1,9 @@
+interface IContext{
+	execute(param: object): boolean;
```
Comment (klumba12) [inline src/core/infrastructure/command.d.ts]: execute can return any

**Review by klumba12 — APPROVED** [review]

**Pull request merged**